lib/plenario/changesets/meta_changesets.ex

```diff
@@ -1,212 +1,182 @@
+  def update(instance, params) do
+    instance
+    |> cast(params, @update_keys)
+    |> validate_required(@required_keys)
     |> unique_constraint(:source_url)
+    |> unique_constraint(:name)
+    |> cast_assoc(:user)
+    |> validate_state()
+    |> test_source_url()
     |> validate_source_type()
-  end
-
-  @doc """
-  Creates a changeset for updating a Meta's descriptive fields
-  """
-  @spec update_description_info(
-          meta :: Meta,
-          params :: %{description: String.t(), attribution: String.t()}
-        ) :: Ecto.Changeset.t()
-  def update_description_info(meta, params) do
-    meta
-    |> cast(params, [:description, :attribution])
-  end
-
-  @doc """
-  Creates a changeset for updating a Meta's refresh information.
-
-  Params include:
-
-    - refresh_rate
-    - refresh_interval
-    - refresh_starts_on
-    - refresh_ends_on
-  """
-  @spec update_refresh_info(meta :: Meta, params :: %{}) :: Ecto.Changeset.t()
-  def update_refresh_info(meta, params) do
-    meta
-    |> cast(params, [:refresh_rate, :refresh_interval, :refresh_starts_on, :refresh_ends_on])
     |> validate_refresh_rate()
     |> validate_refresh_interval()
     |> validate_refresh_ends_on()
+    |> set_slug()
   end
 
-  #  def update_bbox(meta, params) do
-  #    meta
-  #    |> cast(params, [:bbox])
-  #  end
-
-  #  def update_timerange(meta, params) do
-  #    meta
-  #    |> cast(params, [:timerange])
-  #  end
-
-  @doc """
-  Creates a changeset for updating a Meta's next refresh field. This is intended to
-  only be called by automated internal processes, mostly run after an EtlJob.
-  """
-  @spec update_next_refresh(meta :: Meta, params :: %{next_refresh: %DateTime{}}) ::
-          Ecto.Changeset.t()
-  def update_next_refresh(meta, params) do
-    meta
-    |> cast(params, [:next_refresh])
+  defp validate_state(%Ecto.Changeset{valid?: true, changes: changes} = changeset) do
+    case get_field(changeset, :state) do
+      "new" -> changeset
+      _ ->
+        change_keys = MapSet.new(Map.keys(changes))
+        case MapSet.size(MapSet.difference(change_keys, @acceptible_post_new_update_keys)) > 0 do
+          false -> changeset
+          true -> add_error(changeset, :base, "Cannot edit a data set once it's in process.")
+        end
+    end
   end
-
-  # Sets the slug for the meta entry
-  defp set_slug(changeset), do: changeset |> put_change(:slug, Tokens.generate_token())
-
-  # TODO: create some module level attribute for these values
-  defp validate_refresh_rate(changeset) do
-    rr = get_field(changeset, :refresh_rate)
-
-    if Enum.member?(Meta.get_refresh_rate_values(), rr) do
-      changeset
-    else
-      changeset |> add_error(:refresh_rate, "Invalid value `#{rr}`")
+  defp validate_state(changeset), do: changeset
+
+  defp test_source_url(%Ecto.Changeset{valid?: true, changes: %{source_url: source_url}} = changeset) do
+    case HTTPoison.options(source_url) do
```

Comment:
I don't think this is a good idea to test if a page exists. From what I've read not many websites bother to respond properly to an OPTIONS request. You might get a lot of false negatives. Is it not possible to inspect the status code of a GET or POST without waiting for the response body?